File: a2-starting-code/starWar.py
"""
* CSCI3180 Principles of Programming Languages
*
* --- Declaration ---
* For all submitted files, including the source code files and the written
* report, for this assignment:
*
* I declare that the assignment here submitted is original except for source
* materials explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
*
*
* Source material acknowledgements (if any):
*
* Students whom I have discussed with (if any):
"""

import time
import random

# constants
GAME_MAP_ROWS = 16
GAME_MAP_COLS = 48

ENEMY_LIFE = 5
ENEMY_SHOOT_INTERVAL = 5
ENEMY_SIZE_ROWS = 3
ENEMY_SIZE_COLS = 7
ENEMY_SPEED = 1

PLAYER_LIFE = 2
PLAYER_SHOOT_INTERVAL = 5
PLAYER_SIZE_ROWS = 2
PLAYER_SIZE_COLS = 5
PLAYER_SPEED = 1

# enums
class DirectionType:
    DIR_UP = 0
    DIR_DOWN = 1
    DIR_LEFT = 2
    DIR_RIGHT = 3
    DIR_ERROR = 4

class BulletType:
    BUL_FROM_ENEMY = 0
    BUL_FROM_PLAYER = 1

class GameOverType:
    PLAYER_WIN = 1
    ENEMY_WIN = 2

# global variables
global_directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
global_game_map = [[' ' for _ in range(GAME_MAP_COLS)] for _ in range(GAME_MAP_ROWS)]
global_player = None
global_enemy = None
global_bullets = []

class Plane:
    def __init__(self, init_location, init_symbol, init_size, init_speed, init_direction, init_life):
        self._life = init_life
        self._location = init_location
        self._size = init_size
        self._symbol = init_symbol
        self._speed = init_speed
        self._direction = init_direction

    # TODO: add Property Decorator for attributes self._life, self._symbol, self._direction here, including setter.
    @property
    def life(self):
        return self._life
    
    @life.setter
    def life(self, new_life):
        if new_life < 0:
            print("The life value can not be negative")
        else:
            self._life = new_life
    
    @property
    def symbol(self):
        return self._symbol
    
    @symbol.setter
    def symbol(self, new_symbol):
        self._symbol = new_symbol
    
    @property
    def direction(self):
        return self._direction
    
    @direction.setter
    def direction(self, new_direction):
        if new_direction >= DirectionType.DIR_ERROR:
            print("The direction value can only be less than 4")
        else:
            self._direction = new_direction
    
    def draw(self):
        for i in range(self._size[0]):
            for j in range(self._size[1]):
                global_game_map[(self._location[0] + i) % GAME_MAP_ROWS][(self._location[1] + j) % GAME_MAP_COLS] = self._symbol

    def move(self):
        # TODO: Implement the logic of the next position the plane moves to here.
        new_location = ((self._location[0] + (global_directions[self.direction][0] * self._speed + GAME_MAP_ROWS)) % GAME_MAP_ROWS,
                        (self._location[1] + (global_directions[self.direction][1] * self._speed + GAME_MAP_COLS)) % GAME_MAP_COLS)
        self._location = new_location

    def is_collision(self, bullet_location):
        # TODO: Implement the logic to determine whether the plane has been hit by a bullet here.
        check_x = bullet_location[0] >= self._location[0] and bullet_location[0] < (self._location[0] + self._size[0])
        check_y = bullet_location[1] >= self._location[1] and bullet_location[1] < (self._location[1] + self._size[1])
        return check_x and check_y

    def shoot(self, bullet_location, bullet_direction, bullet_type):
        new_bullet = Bullet(bullet_location, self._symbol, bullet_direction, bullet_type)
        global_bullets.append(new_bullet)
    
    def hit(self):
        self.life -= 1
    
class Bullet:
    def __init__(self, init_location, init_symbol, init_direction, init_type):
        assert init_direction == DirectionType.DIR_UP or init_direction == DirectionType.DIR_DOWN
        assert init_type == BulletType.BUL_FROM_ENEMY or init_type == BulletType.BUL_FROM_PLAYER
        self._validity = True
        self._bullet_type = init_type
        self._bullet_on_edge = False
        self._location = init_location
        self._symbol = init_symbol
        self._speed = 1
        self._direction = init_direction

    # TODO: add Property Decorator for attributes self._validity here, including setter.
    @property
    def validity(self):
        return self._validity
    
    @validity.setter
    def validity(self, new_validity):
        self._validity = new_validity
    
    def draw(self):
        if not self._validity:
            return
        global_game_map[self._location[0] % GAME_MAP_ROWS][self._location[1] % GAME_MAP_COLS] = self._symbol

    def move(self):
        if self._bullet_on_edge:
            self.validity = False
            return

        # TODO: Implement the logic of the next position the bullet moves to here.
        new_location = ((self._location[0] + (global_directions[self._direction][0] * self._speed + GAME_MAP_ROWS)) % GAME_MAP_ROWS,
                        (self._location[1] + (global_directions[self._direction][1] * self._speed + GAME_MAP_COLS)) % GAME_MAP_COLS)
        self._location = new_location
        
        # TODO: Implement the logic to determine whether the bullet hits the plane.
        if self._bullet_type == BulletType.BUL_FROM_ENEMY:
            if global_player.is_collision(self._location):
                global_player.hit()
                self._validity = False
        else:
            if global_enemy.is_collision(self._location):
                global_enemy.hit()
                self._validity = False
        
        self.check_on_edge()
    
    def check_on_edge(self):
        # TODO: Implement the logic to determine whether the bullet reaches the map boundary.
        if (self._location[0] <= self._speed) and (self._bullet_type == BulletType.BUL_FROM_PLAYER):
            self._bullet_on_edge = True
        
        if (self._location[0] + self._speed >= GAME_MAP_ROWS) and (self._bullet_type == BulletType.BUL_FROM_ENEMY):
            self._bullet_on_edge = True
        

class Player(Plane):
    def __init__(self, init_location):
        super().__init__(init_location, '^', (PLAYER_SIZE_ROWS, PLAYER_SIZE_COLS), PLAYER_SPEED, DirectionType.DIR_LEFT, PLAYER_LIFE)
        self._shoot_interval = 0
    
    def move(self):
        super().move()
        # TODO: Implement the logic of the player automatically firing a bullet after PLAYER_SHOOT_INTERVAL frames.
        if self._shoot_interval >= PLAYER_SHOOT_INTERVAL:
            self.shoot()
            self._shoot_interval = 0
        else:
            self._shoot_interval += 1

    def shoot(self):
        bullet_location = (self._location[0], self._location[1] + (self._size[1] // 2))
        super().shoot(bullet_location, DirectionType.DIR_UP, BulletType.BUL_FROM_PLAYER)
    
    def speak(self):
        dialogues = ["I'm ready for action!", "No one can defeat me!", "I'll save the world!"]
        print("Player:\t", random.choice(dialogues))
        
    def display_info(self):
        print("Congratulations! You have defeated them.")


class Enemy(Plane):
    def __init__(self, init_location):
        super().__init__(init_location, '$', (ENEMY_SIZE_ROWS, ENEMY_SIZE_COLS), ENEMY_SPEED, DirectionType.DIR_LEFT, ENEMY_LIFE)
        self._shoot_interval = 0

    def move(self):
        super().move()
        # TODO: Implement the logic that enemy has a 10% chance of changing direction in each frame.
        if random.random() * 100 < 10:
            if self.direction == DirectionType.DIR_LEFT:
                self.direction = DirectionType.DIR_RIGHT
            else:
                self.direction = DirectionType.DIR_LEFT
        
        # TODO: Implement the logic of the player automatically firing a bullet after ENEMY_SHOOT_INTERVAL frames.
        if self._shoot_interval >= ENEMY_SHOOT_INTERVAL:
            self.shoot()
            self._shoot_interval = 0
        else:
            self._shoot_interval += 1

    def shoot(self):
        bullet_location = (self._location[0] + self._size[0] - 1, self._location[1] + (self._size[1] // 2))        
        super().shoot(bullet_location, DirectionType.DIR_DOWN, BulletType.BUL_FROM_ENEMY)
    
    def speak(self):
        dialogues = ["You won't escape!", "Prepare to be destroyed!", "I will crush you!"]
        print("Enemy:\t", random.choice(dialogues))
        
    def display_info(self):
        print("Sad! You made sacrifices in this fierce battle.")
    
class Environment:
    def __init__(self):
        self._winner = None
        self._speaker = None
        
        global global_enemy, global_player
        global_enemy = Enemy((0, (GAME_MAP_COLS - ENEMY_SIZE_COLS) // 2))
        global_player = Player((GAME_MAP_ROWS - PLAYER_SIZE_ROWS, (GAME_MAP_COLS - PLAYER_SIZE_COLS) // 2))
        
        # counting down
        print("\n\n\t\tThe game is about to begin!")
        for i in range(3, 0, -1):
            start = time.time()
            while time.time() - start <= 1:
                pass  # 1 s

            if i > 0:
                print("\n\n\t\tCountdown:", i)

    def move_all(self):
        global_player.move()
        global_enemy.move()
        for bullet in global_bullets:
            bullet.move()

    def draw_all(self):
        global global_game_map
        global_game_map = [[0] * GAME_MAP_COLS for _ in range(GAME_MAP_ROWS)]
        global_player.draw()
        global_enemy.draw()
        for obj_ptr in global_bullets:
            obj_ptr.draw()

    def check_state(self):
        to_be_delete = []
        for i in range(len(global_bullets)):
            if not global_bullets[i].validity:
                to_be_delete.append(i)
        for idx in reversed(to_be_delete):
            global_bullets.pop(idx)
        
        # TODO: Implement the logic to determine whether the game should end.
        if global_enemy.life <= 0:
            self._winner = global_player
        if global_player.life <= 0:
            self._winner = global_enemy

    def display_all(self):
        print()
        for _ in range(GAME_MAP_COLS + 2):
            print('-', end='')
        print()
        for i in range(GAME_MAP_ROWS):
            print('|', end='')
            for j in range(GAME_MAP_COLS):
                if global_game_map[i][j] != 0:
                    print(global_game_map[i][j], end='')
                else:
                    print(" ", end='')
            print('|')
        for _ in range(GAME_MAP_COLS + 2):
            print('-', end='')
        print()
        
        # TODO: Implement speaking logic here.
        if random.randint(0, 1) == 0:
            self._speaker = global_player
        else:
            self._speaker = global_enemy
        self._speaker.speak() 
        
        for _ in range(GAME_MAP_COLS + 2):
            print('-', end='')
        print()
        print("ENEMY\tHP: ", global_enemy.life)
        print("PLAYER\tHP: ", global_player.life)

    def display_result(self):
        self._winner.display_info()
    
    def get_input(self):
        print("Please select the moving direction! (w : up, s : down, a: left, d: right)")
        player_input = input()  # get input from player
        if player_input == 'w':
            global_player.direction = DirectionType.DIR_UP
        elif player_input == 's':
            global_player.direction = DirectionType.DIR_DOWN
        elif player_input == 'a':
            global_player.direction = DirectionType.DIR_LEFT
        elif player_input == 'd':
            global_player.direction = DirectionType.DIR_RIGHT

    def run(self):
        self.check_state()
        self.draw_all()
        self.display_all()
        while not self._winner:
            self.get_input()
            self.move_all()
            self.draw_all()
            self.display_all()
            self.check_state()
        self.display_result()
    

if __name__ == "__main__":
    random.seed(555)
    env = Environment()
    env.run()
    

File: a2-starting-code/starWar.cpp
#include <iostream>
#include <ctime>
#include <vector>
#include <cassert>
#include <cstring>
#include <sstream>
#include <chrono>
#include <thread>

class Plane;
class Bullet;
class Player;
class Enemy;

namespace setting
{
    constexpr int gameMapRows = 16;
    constexpr int gameMapCols = 48;

    constexpr int enemyLife = 5;
    constexpr int enemyShootInterval = 5;
    constexpr int enemySizeRows = 3;
    constexpr int enemySizeCols = 7;
    constexpr int enemySpeed = 1;

    constexpr int playerLife = 2;
    constexpr int playerShootInterval = 5;
    constexpr int playerSizeRows = 2;
    constexpr int playerSizeCols = 5;
    constexpr int playerSpeed = 1;
}
enum DirectionType
{
    DIR_UP = 0,
    DIR_DOWN,
    DIR_LEFT,
    DIR_RIGHT,
    DIR_ERROR,
};

enum BulletType
{
    BUL_FROM_ENEMY,
    BUL_FROM_PLAYER,
};

enum GameOverType
{
    PLAYER_WIN = 1,
    ENEMY_WIN,
};

const int global_directions[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
char global_game_map[setting::gameMapRows][setting::gameMapCols];
Player *global_player;
Enemy *global_enemy;
std::vector<Bullet *> global_bullets;

class Plane
{
protected:
    int life; // left HP value
    // for drawing
    std::pair<unsigned int, unsigned int> location;
    std::pair<unsigned int, unsigned int> size;
    char symbol;
    // for moving
    int speed;
    int direction;

public:
    Plane(std::pair<unsigned int, unsigned int> init_location, char init_symbol, std::pair<unsigned int, unsigned int> init_size, int init_speed, int init_direction, int init_life)
    {
        life = init_life;
        location = init_location;
        size = init_size;
        symbol = init_symbol;
        speed = init_speed;
        direction = init_direction;
    }

    int get_life()
    {
        return life;
    }

    void set_life(int new_life)
    {
        if (new_life < 0)
            std::cout << "The life value can not be negative" << std::endl;
        else
            life = new_life;
    }

    char get_symbol()
    {
        return symbol;
    }

    void set_symbol(char new_symbol)
    {
        symbol = new_symbol;
    }

    int get_direction()
    {
        return direction;
    }

    void set_direction(int new_direction)
    {
        if (new_direction >= DIR_ERROR)
            std::cout << "The direction value can only be less than 4" << std::endl;
        else
            direction = new_direction;
    }

    virtual void draw()
    {
        for (int i = 0; i < size.first; i++)
        {
            for (int j = 0; j < size.second; j++)
            {
                global_game_map[(location.first + i) % setting::gameMapRows][(location.second + j) % setting::gameMapCols] = symbol;
            }
        }
    }

    virtual void move()
    {
        location.first = (location.first + (global_directions[direction][0] * speed + setting::gameMapRows)) % setting::gameMapRows;
        location.second = (location.second + (global_directions[direction][1] * speed + setting::gameMapCols)) % setting::gameMapCols;
    }

    bool is_collision(std::pair<int, int> bullet_location)
    {
        bool check_x = bullet_location.first >= location.first && bullet_location.first < (location.first + size.first);
        bool check_y = bullet_location.second >= location.second && bullet_location.second < (location.second + size.second);
        return check_x && check_y;
    }

    virtual void shoot(std::pair<int, int> bullet_location, int bullet_direction, int bullet_type);

    virtual void hit()
    {
        life -= 1;
    }

    virtual void speak() = 0;

    virtual void display_info() = 0;
};

class Bullet
{
private:
    bool validity;
    int bullet_type;
    bool bullet_on_edge;
    // for drawing
    std::pair<unsigned int, unsigned int> location;
    char symbol;
    // for moving
    int speed;
    int direction;

public:
    Bullet(std::pair<unsigned int, unsigned int> init_location, char init_symbol, int init_direction, int init_type)
    {
        assert(init_direction == DIR_UP || init_direction == DIR_DOWN);
        assert(init_type == BUL_FROM_ENEMY || init_type == BUL_FROM_PLAYER);
        validity = true;
        bullet_type = init_type;
        bullet_on_edge = false;
        location = init_location;
        symbol = init_symbol;
        speed = 1;
        direction = init_direction;
    }

    bool get_validity()
    {
        return validity;
    }

    void set_validity(bool new_validity)
    {
        validity = new_validity;
    }

    void draw()
    {
        if (!validity)
            return;
        global_game_map[location.first % setting::gameMapRows][location.second % setting::gameMapCols] = symbol;
    }

    void move();

    void check_on_edge()
    {
        if ((location.first <= speed) && bullet_type == BUL_FROM_PLAYER)
            bullet_on_edge = true;

        if ((location.first + speed >= setting::gameMapRows) && bullet_type == BUL_FROM_ENEMY)
            bullet_on_edge = true;
    }
};

class Player : public Plane
{
private:
    int shoot_interval;

public:
    Player(std::pair<unsigned int, unsigned int> init_location) : Plane(init_location, '^', std::make_pair(setting::playerSizeRows, setting::playerSizeCols), setting::playerSpeed, DIR_LEFT, setting::playerLife)
    {
        shoot_interval = 0;
    }

    void move()
    {
        Plane::move();
        if (shoot_interval++ >= setting::playerShootInterval)
        {
            shoot();
            shoot_interval = 0;
        }
    }

    void shoot();

    void speak()
    {
        std::vector<std::string> dialogues = {"I'm ready for action!", "No one can defeat me!", "I'll save the world!"};
        std::cout << "Player:\t" << dialogues[rand() % dialogues.size()] << std::endl;
    }

    void display_info()
    {
        std::cout << "Congratulations! You have defeated them." << std::endl;
    }
};

class Enemy : public Plane
{
private:
    int shoot_interval;

public:
    Enemy(std::pair<unsigned int, unsigned int> init_location) : Plane(init_location, '$', std::make_pair(setting::enemySizeRows, setting::enemySizeCols), setting::enemySpeed, DIR_LEFT, setting::enemyLife)
    {
        shoot_interval = 0;
    }

    void move()
    {
        Plane::move();
        if ((rand() % 100) < 10)
        {
            if (direction == DIR_LEFT)
                set_direction(DIR_RIGHT);
            else
                set_direction(DIR_LEFT);
        }
        if (shoot_interval++ >= setting::enemyShootInterval)
        {
            shoot();
            shoot_interval = 0;
        }
    }

    void shoot();

    void speak()
    {
        std::vector<std::string> dialogues = {"You won't escape!", "Prepare to be destroyed!", "I will crush you!"};
        std::cout << "Enemy:\t" << dialogues[rand() % dialogues.size()] << std::endl;
    }

    void display_info()
    {
        std::cout << "Sad! You made sacrifices in this fierce battle." << std::endl;
    }
};

void Bullet::move()
{
    if (bullet_on_edge)
    {
        set_validity(false);
        return;
    }

    location.first = (location.first + (global_directions[direction][0] * speed + setting::gameMapRows)) % setting::gameMapRows;
    location.second = (location.second + (global_directions[direction][1] * speed + setting::gameMapCols)) % setting::gameMapCols;

    if (bullet_type == BUL_FROM_ENEMY)
    {
        if (global_player->is_collision(location))
        {
            global_player->hit();
            set_validity(false);
        }
    }
    else
    {
        if (global_enemy->is_collision(location))
        {
            global_enemy->hit();
            set_validity(false);
        }
    }
    check_on_edge();
}

void Plane::shoot(std::pair<int, int> bullet_location, int bullet_direction, int bullet_type)
{
    Bullet *new_bullet = new Bullet(bullet_location, symbol, bullet_direction, bullet_type);
    global_bullets.push_back(new_bullet);
}

void Player::shoot()
{
    std::pair<int, int> bullet_location = location;
    bullet_location.second += (size.second / 2);
    Plane::shoot(bullet_location, DIR_UP, BUL_FROM_PLAYER);
}

void Enemy::shoot()
{
    std::pair<int, int> bullet_location = location;
    bullet_location.first += size.first - 1;
    bullet_location.second += (size.second / 2);
    Plane::shoot(bullet_location, DIR_DOWN, BUL_FROM_ENEMY);
}

class Environment
{
public:
    Plane *winner;
    Plane *speaker;

public:
    Environment()
    {
        global_enemy = new Enemy(std::make_pair(0, (setting::gameMapCols - setting::enemySizeCols) / 2));
        global_player = new Player(std::make_pair(setting::gameMapRows - setting::playerSizeRows, (setting::gameMapCols - setting::playerSizeCols) / 2));
        winner = nullptr;
        speaker = nullptr;

        // counting down
        std::cout << "\n\n\t\tThe game is about to begin!" << std::endl;
        for (int i = 3; i >= 0; i--)
        {
            long start = clock();
            while (clock() - start <= 1000)
                ; // 1 s

            if (i > 0)
                std::cout << "\n\n\t\tCountdown:" << i << std::endl;
        }
    }

    void move_all()
    {
        global_player->move();
        global_enemy->move();
        for (int i = 0; i < global_bullets.size(); i++)
        {
            global_bullets[i]->move();
        }
    }

    void draw_all()
    {
        memset(global_game_map, 0, sizeof(global_game_map));
        global_player->draw();
        global_enemy->draw();
        for (int i = 0; i < global_bullets.size(); i++)
        {
            global_bullets[i]->draw();
        }
    }

    void check_state()
    {
        std::vector<int> to_be_delete;
        for (int i = 0; i < global_bullets.size(); i++)
        {
            if (global_bullets[i]->get_validity() == false)
            {
                to_be_delete.push_back(i);
            }
        }
        for (auto it = to_be_delete.rbegin(); it != to_be_delete.rend(); ++it)
        {
            global_bullets.erase(global_bullets.begin() + *it);
        }

        if (global_enemy->get_life() <= 0)
            winner = global_player;
        if (global_player->get_life() <= 0)
            winner = global_enemy;
    }

    void display_all()
    {
        std::cout << std::endl;
        for (int j = 0; j < setting::gameMapCols + 2; j++)
        {
            std::cout << '-';
        }
        std::cout << std::endl;
        for (int i = 0; i < setting::gameMapRows; i++)
        {
            std::cout << '|';
            for (int j = 0; j < setting::gameMapCols; j++)
            {
                if (global_game_map[i][j] != 0)
                    std::cout << global_game_map[i][j];
                else
                    std::cout << " ";
            }
            std::cout << '|';
            std::cout << std::endl;
        }
        for (int j = 0; j < setting::gameMapCols + 2; j++)
        {
            std::cout << '-';
        }
        std::cout << std::endl;

        if ((rand() % 2) == 0) 
        {
            speaker = global_player;
        }
        else
        {
            speaker = global_enemy;
        }
        speaker->speak();

        for (int j = 0; j < setting::gameMapCols + 2; j++)
        {
            std::cout << '-';
        }
        std::cout << std::endl;
        std::cout << "ENEMY\tHP: "
                  << global_enemy->get_life() << std::endl;
        std::cout << "PLAYER\tHP: "
                  << global_player->get_life() << std::endl;
    }

    void display_result()
    {
        winner->display_info();
    }

    void get_input()
    {
        std::cout << "Please select the moving direction! (w : up, s : down, a: left, d: right)" << std::endl;
        char input;
        std::cin >> input; // get input from player
        if (input == 'w')
            global_player->set_direction(DIR_UP);
        else if (input == 's')
            global_player->set_direction(DIR_DOWN);
        else if (input == 'a')
            global_player->set_direction(DIR_LEFT);
        else if (input == 'd')
            global_player->set_direction(DIR_RIGHT);
    }

    void run()
    {
        check_state();
        draw_all();
        display_all();
        while (!winner)
        {
            get_input();
            move_all();
            draw_all();
            display_all();
            check_state();
        }
        display_result();
    }
};

int main()
{
    srand(555);
    Environment env;
    env.run();
}
